### PR TITLE
Feature/1385 update autodiff tests

### PR DESCRIPTION
**Pull request by bob-carpenter**
## Summary

Consolidates remaining fwd and rev tests into mix for all scalar functions of 2 arguments or fewer.   Full list below.  The remaining functions do not have derivatives.

*Warning:* Branches from `feature/1382-remove-fvar-nan-checks`, so PR: #1384 should be reviewed and merged before this one is merged.

#### Consolidated tests

The following tests were consolidated from `rev` and `fwd` into `mix` using the test framework.  Test cases were carried over.

* [x] `{fwd,rev}/scal/fun/abs_test.cpp` -> `mix/scal/fun/abs_test.cpp`
* [x] `{fwd,rev}/scal/fun/atan2_test.cpp` -> `mix/scal/fun/atan2_test.cpp`
* [x] `{fwd,rev}/scal/fun/bessel_first_kind_test.cpp` -> `mix/scal/fun/bessel_first_kind_test.cpp`
* [x] `{fwd,rev}/scal/fun/bessel_second_kind_test.cpp` -> `mix/scal/fun/bessel_secon_kind_test.cpp`
* [x] `{fwd,rev}/scal/fun/beta_test.cpp` -> `mix/scal/fun/beta_test.cpp`
* [x] `{fwd,rev}/scal/fun/binary_log_loss_test.cpp` -> `mix/scal/fun/binary_log_loss_test.cpp`
* [x] `{fwd,rev}/scal/fun/binomial_coefficient_log_test.cpp` -> `mix/scal/fun/binomial_coefficient_log_test.cpp`
* [x] `{fwd,rev}/scal/fun/falling_factorial_test.cpp` -> `mix/scal/fun/falling_factorial_test.cpp`
* [x] `{fwd,rev}/scal/fun/fdim_test.cpp` -> `mix/scal/fun/fdim_test.cpp`
* [x] `{fwd,rev}/scal/fun/fmax_test.cpp` -> `mix/scal/fun/fmax_test.cpp`
* [x] `{fwd,rev}/scal/fun/fmin_test.cpp` -> `mix/scal/fun/fmin_test.cpp`
* [x] `{fwd,rev}/scal/fun/fmod_test.cpp` -> `mix/scal/fun/fmod_test.cpp`
* [x] `{fwd,rev}/scal/fun/gamma_p_test.cpp` -> `mix/scal/fun/gamma_p_test.cpp`
* [x] `{fwd,rev}/scal/fun/gamma_q_test.cpp` -> `mix/scal/fun/gamma_q_test.cpp`
* [x] `{fwd,rev}/scal/fun/hypot_test.cpp` -> `mix/scal/fun/hypot_test.cpp`
* [x] `{fwd,rev}/scal/fun/lbeta_test.cpp` -> `mix/scal/fun/lbeta_test.cpp`
* [x] `{fwd,rev}/scal/fun/lmgamma_test.cpp` -> `mix/scal/fun/lmgamma_test.cpp`
* [x] `{fwd,rev}/scal/fun/log_diff_exp_test.cpp` -> `mix/scal/fun/log_diff_exp_test.cpp`
* [x] `{fwd,rev}/scal/fun/log_falling_factorial_test.cpp` -> `mix/scal/fun/log_falling_factorial_test.cpp`
* [x] `{fwd,rev}/scal/fun/log_rising_factorial_test.cpp` -> `mix/scal/fun/log_rising_factorial_test.cpp`
* [x] `{fwd,rev}/scal/fun/log_inv_logit_diff_test.cpp` -> `mix/scal/fun/log_inv_logit_diff_test.cpp`
* [x] `{fwd,rev}/scal/fun/log_sum_exp_test.cpp` -> `mix/scal/fun/log_sum_exp_test.cpp`
* [x] `{fwd,rev}/scal/fun/modified_bessel_first_kind_test.cpp` -> `mix/scal/fun/modified_bessel_first_kind_test.cpp`
* [x] `{fwd,rev}/scal/fun/modified_bessel_second_kind_test.cpp` -> `mix/scal/fun/modified_bessel_second_kind_test.cpp`
* [x] `{fwd,rev}/scal/fun/multiply_log_test.cpp` -> `mix/scal/fun/multiply_log_test.cpp`
* [x] `{fwd,rev}/scal/fun/owens_t_test.cpp` -> `mix/scal/fun/owens_t_test.cpp`
* [x] `{fwd,rev}/scal/fun/pow_test.cpp` -> `mix/scal/fun/pow_test.cpp`
* [x] `{fwd,rev}/scal/fun/rising_factorial_test.cpp` -> `mix/scal/fun/rising_factorial_test.cpp`
* [x] `{fwd,rev}/scal/fun/sign_test.cpp` -> `mix/scal/fun/sign_test.cpp`

The operator tests in the core are now part of the mix core tests using the test framework.

* [x] `fwd/core/fvar_test.cpp` -> `mix/core/operator_*`

The following had reverse-mode tests, but no forward or mixed-mode tests.  There are now mixed tests (and forward-mode tests) using the framework.

* [x] `rev/scal/fun/sign_test.cpp` -> `mix/scal/fun/squared_distance_test.cpp` [no pre-existing fwd or mix tests]
* [x] `rev/scal/fun/step_test.cpp` -> `mix/scal/fun/squared_distance_test.cpp` [no pre-existing fwd or mix tests]

#### Notes

* removed buggy implementation of `binomial_coefficient_log` in forward mode and just went with working templated version
* replaced direct includes from boost with stan::math includes where found
* fixed bug in forward mode autodiff for one function, but can't remember which one---it's the one with  a factor removed in one of its forward-mode overloads

Describe the contents of the pull request. The issue describes the problem that needs to be fixed, whereas the text here should help the reviewer figure out to what extent this pull request solves the issue.

Describe implementation details and any relevant information that would help the code reviewer understand this pull request. If there is anything special you need to draw someone's attention to, do it here. Link to any related pull requests or Discourse threads.

## Tests

It's all tests.

## Side Effects

No.

## Checklist

- [x] Math issue #1385

- [x] Copyright holder: (fill in copyright holder information)

Columbia University

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by SteveBronder** [comment]
Is this ready for review? I know this is normally in @bbbales2 area but if he doesn't have time I can look this over

**Comment by wds15** [comment]
WHOW: +700 ... -11000 lines of code!

+1

**Comment by bbbales2** [comment]
@SteveBronder I do have time, but if you want at it have at it :D.

**Comment by bbbales2** [comment]
@SteveBronder did you have a look at this? I'll do it tomorrow if not.

@bob-carpenter looks like there are some merge conflicts (but they look easy for what it's worth)

**Comment by bob-carpenter** [comment]
I'll fix the merge conflicts and push.

**Comment by SteveBronder** [comment]
Apologies! Flying back to NYC today I can look at this tmrw

**Comment by bbbales2** [comment]
@SteveBronder Nah nah you're good. Seeyah back in NYC. I got it. Just wanted to avoid accidental duplicate work.

File: test/unit/math/mix/scal/fun/abs_test.cpp [hunk 1]
```diff
@@ -1,101 +1,18 @@
...
-  test_nan_mix(abs_, false);
+#include <test/unit/math/test_ad.hpp>
+
+TEST(mixScalFun, abs) {
+  auto f = [](const auto& x) { return stan::math::abs(x); };
+  stan::test::expect_common_nonzero_unary(f);
+  stan::test::expect_value(f, 0);
+  stan::test::expect_value(f, 0.0);
+
+  stan::test::expect_ad(f, -3);
```
Comment (bbbales2) [inline test/unit/math/mix/scal/fun/abs_test.cpp]: Isn't this function vectorized? Was there a special mechanism for testing vectorized functions?

File: test/unit/math/mix/scal/fun/falling_factorial_test.cpp [hunk 2]
```diff
@@ -1,158 +1,22 @@
...
-  fvar<var> a(5.0, 1.0);
-  fvar<var> c = falling_factorial(a, 3);
+  stan::test::expect_ad(f(3), 5);
 
-  EXPECT_FLOAT_EQ(falling_factorial(5, 3), c.val_.val());
-  EXPECT_FLOAT_EQ(
-      falling_factorial(5, 3) * (digamma(5 + 1) - digamma(5 + 1 - 3)),
-      c.d_.val());
+  // 3rd order derivatives near zero
+  stan::test::ad_tolerances tols;
```
Comment (bbbales2) [inline test/unit/math/mix/scal/fun/falling_factorial_test.cpp]: Do you think we should keep a list somewhere of functions that require changes to the finite difference tolerances to test?

Like in an issue?

File: test/unit/math/mix/scal/fun/gamma_q_test.cpp [hunk 3]
```diff
@@ -1,354 +1,19 @@
...
+  stan::test::expect_ad(f, 0.5001, 1.0001);
+  stan::test::expect_ad(f, 0.5, 1.0);
+  stan::test::expect_ad(f, 1.0, 1.0);
+
+  double nan = std::numeric_limits<double>::quiet_NaN();
+  stan::test::expect_ad(f, 1.0, nan);
+  stan::test::expect_ad(f, nan, 1.0);
+  stan::test::expect_ad(f, nan, nan);
+
+  // this still fails forward mode; left regression test in rev/scal/fun
```
Comment (bbbales2) [inline test/unit/math/mix/scal/fun/gamma_q_test.cpp]: If we decide to document these, here's one.

File: test/unit/math/mix/scal/fun/log_diff_exp_test.cpp [hunk 4]
```diff
@@ -1,358 +1,29 @@
...
+  // stan::test::expect_common_nonzero_binary(f);
+  stan::test::expect_ad(f, 1.2, 0.5);
+  stan::test::expect_ad(f, 5.0, 2.0);
+  stan::test::expect_ad(f, 9.0, 6.0);
+  stan::test::expect_ad(f, 0.1, 0.0);
+  stan::test::expect_ad(f, 3.01, 2.0);
+  stan::test::expect_ad(f, 2.0, 1.0);
+
+  // hessians near zero have high relative error but small absolute error
+  stan::test::ad_tolerances tols;
```
Comment (bbbales2) [inline test/unit/math/mix/scal/fun/log_diff_exp_test.cpp]: If we decide to document these, here's one.

File: test/unit/math/mix/scal/fun/log_sum_exp_test.cpp [hunk 5]
```diff
@@ -1,354 +1,40 @@
...
+  stan::test::expect_ad(f, 1.0, 2.0);
+  stan::test::expect_ad(f, 1.4, 0.5);
+  stan::test::expect_ad(f, 1.2, 0.6);
+  stan::test::expect_ad(f, 2.0, 1.0);
+  stan::test::expect_ad(f, 3.0, 5.0);
+  stan::test::expect_ad(f, 3.0, 6.0);
+  stan::test::expect_ad(f, 3.4, 0.9);
+  stan::test::expect_ad(f, 5.0, 2.0);
+
+  stan::test::ad_tolerances tols;
```
Comment (bbbales2) [inline test/unit/math/mix/scal/fun/log_sum_exp_test.cpp]: If we decide to document these, here's one.

File: test/unit/math/mix/scal/fun/rising_factorial_test.cpp [hunk 6]
```diff
@@ -1,155 +1,23 @@
...
-  fvar<var> a(5.0, 1.0);
-  fvar<var> c = rising_factorial(a, 3);
+  stan::test::expect_ad(f(1), 4.0);
+  stan::test::expect_ad(f(3), 5.0);
+  stan::test::expect_ad(f(4), 4.0);
 
-  EXPECT_FLOAT_EQ(rising_factorial(5, 3), c.val_.val());
-  EXPECT_FLOAT_EQ(rising_factorial(5, 3) * (digamma(5 + 3) - digamma(5)),
-                  c.d_.val());
+  // 3rd derivatives close to zero and rel tolerance fails
```
Comment (bbbales2) [inline test/unit/math/mix/scal/fun/rising_factorial_test.cpp]: If we decide to document these, here's one.

File: test/unit/math/fwd/core/fvar_test.cpp [hunk 7]
```diff
@@ -25,40 +27,9 @@ TEST(mathFwdCoreFvar, ctor) {
...
-  EXPECT_FLOAT_EQ(3.07, d.val_);
-
-  EXPECT_FLOAT_EQ(3.07, (d++).val_);
-  EXPECT_FLOAT_EQ(4.07, d.val_);
-
-  EXPECT_FLOAT_EQ(5.07, (++d).val_);
-  EXPECT_FLOAT_EQ(5.07, d.val_);
-
-  double nan = std::numeric_limits<double>::quiet_NaN();
-  fvar<double> f(nan);
```
Comment (bbbales2) [inline test/unit/math/fwd/core/fvar_test.cpp]: Isn't this the only test that needs removed? Or do all the other tests just seem extraneous?

File: test/unit/math/mix/scal/fun/log_diff_exp_test.cpp (hunk 4 above)
Comment (SteveBronder) [reply test/unit/math/mix/scal/fun/log_diff_exp_test.cpp]: If we want to doc these may be good to add to the autodiff page of the new stan math site I'm setting up

https://github.com/stan-dev/math/pull/1397

**Comment by bob-carpenter** [comment]
> Isn't that implicitly assuming the derivative is zero?

No.  In the tests, when computing gradients and Hessians with `fvar`, we'll set each input's variable to 1 in turn.  That'll test all the derivative propagation.

> Regarding the merge conflicts. Those were changes from another thread, right? 

They were improvements in the constructors for `fvar` that Rok suggested in the fvar change.

> behavior of the gradients when the value is NaN is now undefined?

They're defined by the functions the way all gradients are defined.  Before, we had them automatically become NaN if the value is NaN.  For example, NaN + NaN will have values of 1 for the gradients as that's the way derivatives work for addition, whereas NaN * NaN will have values of NaN for the gradients.

File: test/unit/math/mix/scal/fun/falling_factorial_test.cpp (hunk 2 above)
Comment (bob-carpenter) [reply test/unit/math/mix/scal/fun/falling_factorial_test.cpp]: I don't see any reason to.  Most functions would require changes to tolerances if values got more extreme.  We could try to evaluate analytical changes here, but really it's a problem with the way the framework does relative value comparison for values close to zero.  It's a hard problem in general.

File: test/unit/math/fwd/core/fvar_test.cpp (hunk 7 above)
Comment (bob-carpenter) [reply test/unit/math/fwd/core/fvar_test.cpp]: I'm trying to reduce tests to just testing inputs with finite differences.  Was there a test you think is missing?  

I could just leave all the old test code, but that seems confusing.

File: test/unit/math/mix/scal/fun/falling_factorial_test.cpp (hunk 2 above)
Comment (bbbales2) [reply test/unit/math/mix/scal/fun/falling_factorial_test.cpp]: Cool. They'll be easy to find if we ever need to via grepping.

File: test/unit/math/fwd/core/fvar_test.cpp (hunk 7 above)
Comment (bbbales2) [reply test/unit/math/fwd/core/fvar_test.cpp]: It seems to be testing operator-, operator+=, operator++ or whatnot? If these are tested elsewhere that's cool

**Comment by bob-carpenter** [comment]
Yes, they're

...

> In test/unit/math/fwd/core/fvar_test.cpp:
> ...
> -  EXPECT_FLOAT_EQ(5.07, (++d).val_);
> -  EXPECT_FLOAT_EQ(5.07, d.val_);
> -
> -  double nan = std::numeric_limits<double>::quiet_NaN();
> -  fvar<double> f(nan);
> 
> It seems to be testing operator-, operator+=, operator++ or whatnot? If these are tested elsewhere that's cool

Yes, they're tested much more extensively in mix/core using the test framework.  I create functors to test before and after values at all levels of autodiff plus derivatives.  That was in the original framework PR.  Now I'm going through and cleaning up redundant tests.

File: test/unit/math/mix/scal/fun/abs_test.cpp (hunk 1 above)
Comment (bbbales2) [reply test/unit/math/mix/scal/fun/abs_test.cpp]: I super-apologize cause I thought you responded to this and now I'm not finding the response. I vaguely remember you saying this is covered (presumably there's extensive testing of the mechanism that does the vectorization). Was that is?

**Comment by bbbales2** [comment]
I think the only outstanding comment worth responding to is https://github.com/stan-dev/math/pull/1393#discussion_r334291891 (and I vaguely remember you responded to this already -- or at least I thought you did). Other than that this is good to go. I'll approve it after the release is packaged just so it doesn't accidentally get merged.

File: test/unit/math/mix/scal/fun/abs_test.cpp (hunk 1 above)
Comment (bob-carpenter) [reply test/unit/math/mix/scal/fun/abs_test.cpp]: No need to apologize at all!  No, I forgot to respond.  `abs` is not vectorized.  It was deprecated in favor of `fabs`.  We may want to change that now that we're better at type infererence in C++---it used to be a struggle just to get things to compile with these functions that were in some top-level namespaces and not others and had return-type dependendent signatures (int returns int, double returns double).

And no rush on merging this---it's just testing.

**Comment by bbbales2** [comment]
@seantalts This is good to go. This has been sitting for a few days -- do we need to remerge develop into this before merging this into develop? If not go ahead and press the pretty green merge button.

**Pull request merged**